src/runtime_src/core/tools/common/ReportAsyncError.cpp

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright (C) 2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include <ctime>
+#include "ReportAsyncError.h"
+#include "core/common/query_requests.h"
+#include "core/common/device.h"
+#include "core/include/xrt.h"
+#include "core/include/experimental/xrt_error.h"
+#include "core/include/xrt_error_code.h"
+#include <boost/algorithm/string.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/date_time/posix_time/posix_time_io.hpp>
+namespace qr = xrt_core::query;
+
+boost::property_tree::ptree
+populate_async_error(const xrt_core::device * device)
+{
+  boost::property_tree::ptree pt;
+  boost::property_tree::ptree error_array;
+  auto dhdl = xrtDeviceOpenFromXcl(device->get_device_handle());
+  for (xrtErrorClass ecl = XRT_ERROR_CLASS_FIRST_ENTRY; ecl < XRT_ERROR_CLASS_UNKNOWN; ecl = xrtErrorClass(ecl+1)) {
+    xrtErrorCode errorCode = 0;
+    uint64_t timestamp = 0;
+    int rval = xrtErrorGetLast(dhdl, ecl, &errorCode, &timestamp);
+    /**
+     * In case of no error for given class, errorCode and timestamp will be zero,
+     * but zero errorCode is valid error, so checking for only timestamp.
+     */
+    if (rval == 0 && errorCode && timestamp) {
+      size_t len = 0;
+      if (xrtErrorGetJson(dhdl, errorCode, nullptr, 0, &len))

```

Comment:
Please refactor this code and call the method:
```
void
error_code_to_json(xrtErrorCode ecode, boost::property_tree:ptree & pt)

```instead

Reply:
done